notebooks/03-query/15-pivot-unpivot.ipynb

```diff
@@ -0,0 +1,372 @@
+    "```\n",
+    "- It lists a set of aggregations to calculate: `SUM(added) as added, SUM(deleted) as deleted`\n",
+    "- The FOR clause indicated the pivot column: `\"isRobot\"`\n",
+    "- The IN clause lists the pivot values of the column and what to call them: `IN ('true' as robot, 'false' as human)`\n",
+    "\n",
+    "The resulting columns from this operation are a combination in the form \\<pivot_value_name\\>__\\<aggregation_name\\>. So in this case the result pivot columns will be:\n",
+    "```\n",
+    "\"robot_added\"\n",
+    "\"human_added\"\n",
+    "\"robot_deleted\"\n",
+    "\"human_deleted\"\n",
+    "```\n",
+    "which can now form part of the SELECT clause in the query.\n",
+    "\n",
+    "\n",
+    "The following query uses the results of the pivot to then aggregate totals for the pivot columns by channel. The aggregations being calculated could be NULL in the case where \"isRobot\"='true' (or 'false') does not exist for a given channel. The query calculates a ratio of human to total changes for each channel and the result is sorted on this ratio such that the channels with the highest proportion of robot updates are listed first. So in order to calculate this metric even in presence of NULLs the query uses COALESCE on all the pivoted metrics.\n",
+    "\n",
+    "The result is much cleaner than the prior query and it is easy to determine which channels have the most robotic activity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "65c489a8-a7f2-4cc0-b30c-f30d8ed32be2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT \"channel\", \n",
+    "        SUM(\"robot_added\")       AS \"added_robot\", \n",
+    "        SUM(\"robot_deleted\")     AS \"deleted_robot\", \n",
+    "        SUM(\"human_added\")       AS \"added_human\",\n",
+    "        SUM(\"human_deleted\")     AS \"deleted_human\", \n",
+    "        \n",
+    "        SAFE_DIVIDE( SUM(\"human_deleted\" + \"human_added\") * 1.0 , \n",
+    "                     SUM( \"robot_deleted\" + \"robot_added\" + \"human_deleted\" + \"human_added\" )\n",
+    "                     \n",
+    "                    )            AS \"human_ratio\"\n",
+    "                    \n",
+    "FROM\n",
+    "(\n",
+    "    SELECT\n",
+    "      \"channel\", \n",
+    "       COALESCE(\"robot_added\",0) AS \"robot_added\", \n",
+    "       COALESCE(\"human_added\",0) AS \"human_added\", \n",
+    "       COALESCE(\"robot_deleted\",0) AS \"robot_deleted\", \n",
+    "       COALESCE(\"human_deleted\",0) AS \"human_deleted\"\n",
+    "    FROM \"example-pivot-unpivot\" \n",
+    "    \n",
+    "    PIVOT ( SUM(added) as added, \n",
+    "            SUM(deleted) as deleted \n",
+    "            \n",
+    "            FOR \"isRobot\" IN ('true' as robot, 'false' as human)\n",
+    "          )\n",
+    ")x\n",
+    "GROUP BY 1\n",
+    "ORDER BY 6\n",
+    "'''\n",
+    "display_client.sql(sql)"
```

Comment:
added a LIMIT in this one because it is sorted to show the most egregious robotic managed channels.
and added commentary on the fact that the 20 rows shown are the ones with the highest proportion of robot updates.